editor/src/core/shared/element-path-tree.ts

```diff
@@ -133,6 +134,73 @@ export function reorderTree(trees: ElementPathTrees, metadata: ElementInstanceMe
   }
 }
 
+function maybeReorderDynamicChildren(
+  children: ElementPathTree[],
+  metadata: ElementInstanceMetadataMap,
+): ElementPathTree[] {
+  function keepOnlyUnique(path: ElementPath, index: number, array: ElementPath[]) {
+    return array.indexOf(path) === index
+  }
+
+  // Get all the dynamic children and sort them
+  // alphabetically, since their paths will have incremental indexes (~~~N).
+  const dynamicChildren = children
+    .filter((child) => EP.isDynamicPath(child.path))
+    .sort((a, b) => a.pathString.localeCompare(b.pathString))
+
+  // If there are no dynamic children, do nothing.
+  if (dynamicChildren.length === 0) {
+    return children
+  }
+
+  const metadataKeys = Object.keys(metadata)
+
+  // Build a stack to reorder dynamic children
+  const dynamicChildrenStack = dynamicChildren
+    .sort((a, b) => {
+      return metadataKeys.indexOf(a.pathString) - metadataKeys.indexOf(b.pathString)
+    })
+    .map((c) => EP.dynamicPathToStaticPath(c.path))
+    .filter(keepOnlyUnique)
+
+  // Group the children paths based on their non-dynamic paths.
+  // E.g.:
+  //  +--------------+     +-----------+
+  //  | foo/bar      |     | foo/bar   |
+  //  | foo/baz~~~1  |  →  | foo/baz   |
+  //  | foo/baz~~~2  |     | foo/qux   |
+  //  | foo/baz~~~3  |     | foo/corge |
+  //  | foo/qux      |     +-----------+
+  //  | foo/corge~~1 |
+  //  +--------------+
+  let groupedPaths: ElementPath[] = []
+  for (const child of children) {
+    if (EP.isDynamicPath(child.path)) {
+      const nextDynamicChild = dynamicChildrenStack.shift()
+      if (nextDynamicChild != null) {
+        groupedPaths.push(nextDynamicChild)
+      }
+    } else {
+      groupedPaths.push(child.path)
+    }
+  }
+
+  // Now calculate the complete full children array,
+  // expanding the grouped paths and replacing them with the
+  // list of sorted dynamic paths calculated earlier.
+  const expandedChildren = groupedPaths.flatMap((path): ElementPathTree[] => {
+    const matchingDynamicChildren = dynamicChildren.filter(
+      (other) => EP.dynamicPathToStaticPath(other.path) === path,
```

Comment:
I think this should also use `EP.pathsEqual` we should not rely on reference equality for element paths.

Reply:
https://github.com/concrete-utopia/utopia/pull/3811/commits/96eb805a1f7aa28c783f6ba30d1e50a05b4bf9cc